testplan/runnable/base.py

```diff
@@ -393,106 +394,130 @@ def schedule(self, task=None, resource=None, uid=None, **options):
-        if uid in self._tests:
-            self.logger.error(
-                "Skip adding {} with uid {}.. already added.".format(
-                    runnable, uid
-                )
+        local_runner = self.resources.first()
+        resource = resource or local_runner
+
+        if resource not in self.resources:
+            raise RuntimeError(
+                'Resource "{}" does not exist.'.format(resource)
             )
-            return uid
 
-        if isinstance(runnable, Entity):
-            runnable.cfg.parent = self.cfg
+        # The runnable added into TestRunner can be: 1> a real test entity
+        # 2> a task wraps a test entity 3> a callble returns a test entity
+        if isinstance(runnable, Runnable):
             runnable.parent = self
+            runnable.cfg.parent = self.cfg
         elif isinstance(runnable, Task):
             pass
         elif callable(runnable):
-            runnable.parent_cfg = self.cfg
             runnable.parent = self
+            runnable.parent_cfg = self.cfg
 
-        # Check if test should not be added only when a filter is used.
-        if (
-            type(self.cfg.test_filter) is not filtering.Filter
-            or self.cfg.test_lister is not None
-        ):
-            if not self.should_be_added(runnable):
-                return None
+        # Get the real test target (a test entity)
+        target = self._verify_test_target(runnable)
+        if not target:
+            return None
 
-        local_runner = self.resources.first()
-        if resource is None:
-            resource = local_runner
+        uid = target.uid()
+        part = getattr(target.cfg, "part", None)
 
-        if resource not in self.resources:
-            raise RuntimeError(
-                'Resource "{}" does not exist.'.format(resource)
+        # When running interactively, add all real test targets into the local
+        # runner even if they were scheduled into a pool. It greatly simplifies
+        # the interactive runner if it only has to deal with the local runner.
+        if self.cfg.interactive_port is not None:
+            # Only the first Multitest part will be added into TestRunner
+            # and it is considered as a complete test entity
+            if uid in self._tests:
+                if part is None:
+                    self.logger.error(
+                        'Skip adding %s, target uid "%s" already added.',
+                        runnable,
+                        uid,
+                    )
+            else:
+                target.cfg._options["part"] = None
+                self._tests[uid] = local_runner
+                self.resources[local_runner].add(target, uid)
+            return uid
+
+        # In batch mode uid of test target MUST be unique unless it is
+        # a part of Multitest and is wrapped in a task
+        if uid in self._tests and not (isinstance(runnable, Task) and part):
+            self.logger.error(
```

Comment:
I think this is too mild, if user tries to add 2 runnable with the same uid, we shall error out the plan?

Reply:
if user add 2 test entities with the same uid, then error logging. only when they are multitest-parts then TestRunner can accept it, however part should be scheduled as a task.

Reply:
after discussion, the new implementation can add multitest part without wrapped in a task.